src/dmd/typesem.d

```diff
@@ -1298,9 +1298,16 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
             e = e.implicitCastTo(sc, fparam.type);
 
             // default arg must be an lvalue
-            if (isRefOrOut && !isAuto &&
-                !(global.params.previewIn && (fparam.storageClass & STC.in_)))
-                e = e.toLvalue(sc, e);
+            if (isRefOrOut && !isAuto)
+            {
+                if (!(global.params.previewIn && (fparam.storageClass & STC.in_)) &&
+                    !(global.params.rvalueRefParam))
+                {
+                    e = e.toLvalue(sc, e);
+                    if (e.op == TOK.error)
+                        errorSupplemental(e.loc, "use `-preview=in` or `preview=rvaluerefparam`");
```

Comment:
Don't think we should recommend those `-preview` switch as neither have been approved by our BDFLs.

Reply:
Ok with me to removed it.

Reply:
I'd rather leave this in. 

https://github.com/dlang/DIPs/blob/master/DIPs/rejected/DIP1016.md

reject because
```d
void fun(ref int x);
fun(10);
```
which is disallowed by the current implementation. The current implementation was done because @TurkeyMan quite rightly complained that the review process was not good and that there is utility in the feature.

Reply:
Why is

```d
void fun(ref int x);
fun(10);
```

disallowed?

Reply:
https://github.com/dlang/DIPs/blob/master/DIPs/rejected/DIP1016.md#formal-assessment

Reply:
@thewilsonator : I would appreciate if you gave me the time to reply to this. Whether or not you or someone else feels the review process wasn't good doesn't change the fact that the BDFLs decision have been expressed clearly and this is somewhat going against it.

And regarding `-preview=in`, as much as I want to recommend it to people I will not do it through the compiler as long as we haven't reached a consensus on it.

Reply:
I don't understand what's the problem here. For me with `-preview=rvaluerefparam`

```d
void f(ref const int x) {}
void g(const ref shared long x) {}

unittest {
    f(10);
    g(10L);
}
```

fails as

```
foo.d(6,6): Error: function `foo.g(ref shared(const(long)) x)` is not callable using argument types `(long)`
foo.d(6,6):        cannot pass rvalue argument `10L` of type `long` to parameter `ref shared(const(long)) x`
```

. I see no problem with this behaviour.

Reply:
@nordlow : I submitted https://github.com/dlang/dmd/pull/12425
As I explain there, we shouldn't recommend preview any switches which were completely broken until last release, nor should we start recommending switches for which there is no clear plan.

Reply:
I agree. It doesn't seem to be that difficult to adjust the corrnercases for `-refvaluerefparam` and make it activate by default. This is such a cornercase.